### ISSUE TITLE
pnet::datalink::interfaces() returns wrong MAC addresses since nightly-2022-11-23

### ISSUE THREAD

**Issue by maxmouchet** [libpnet/libpnet#594]
Consider the following example:
```rust
use pnet::datalink;

fn main() {
    for interface in pnet::datalink::interfaces() {
        println!("{} {:?}", interface.name, interface.mac);
    }
}
```

This returns "invalid" MAC addresses. These are not the correct MAC addresses for my network interfaces, and they seem to follow a strange pattern (01, 02, 03, ...):
```
lo0 Some(01:00:18:03:00:00)
gif0 Some(02:00:37:04:00:00)
stf0 Some(03:00:39:04:00:00)
anpi1 Some(04:00:06:05:06:00)
anpi0 Some(05:00:06:05:06:00)
en3 Some(06:00:06:03:06:00)
en4 Some(07:00:06:03:06:00)
en1 Some(08:00:06:03:06:00)
en2 Some(09:00:06:03:06:00)
bridge0 Some(0a:00:06:07:06:00)
ap1 Some(0b:00:06:03:06:00)
en0 Some(0c:00:06:03:06:00)
awdl0 Some(0d:00:06:05:06:00)
llw0 Some(0e:00:06:04:06:00)
utun0 Some(0f:00:01:05:00:00)
utun1 Some(10:00:01:05:00:00)
utun2 Some(11:00:01:05:00:00)
utun3 Some(12:00:01:05:00:00)
utun4 Some(13:00:01:05:00:00)
```

This happens with `libpnet 0.31.0` since `nightly-2022-11-23`. `nightly-2022-11-22` and before are good.

Maybe this is due to an upcoming change in Rust that will affect libpnet, or it's just an issue with Rust nightly, in which case feel free to close :)

**Comment by mrmonday** [comment]
Could you please share:

 * Which operating system you are using
 * The version you are using

This is probably a bug in libpnet.

**Comment by maxmouchet** [comment]
This is on macOS 13.1.

**Comment by jahofmann** [comment]
From the timing this could be another struct affected by the problem mentioned in #600.

You can try adding `#[repr(C)]` to https://github.com/libpnet/libpnet/blob/84d6cb6822e9b90232cdd36c8189827c36b0344c/pnet_datalink/src/bindings/bpf.rs#L98

**Comment by maxmouchet** [comment]
That does it, thanks! I can open a PR later on.

```diff
diff --git a/pnet_datalink/src/bindings/bpf.rs b/pnet_datalink/src/bindings/bpf.rs
index 9a938a4..81cfaa0 100644
--- a/pnet_datalink/src/bindings/bpf.rs
+++ b/pnet_datalink/src/bindings/bpf.rs
@@ -95,6 +95,7 @@ pub struct ifreq {
     target_os = "macos",
     target_os = "ios"
 ))]
+#[repr(C)]
 pub struct sockaddr_dl {
     pub sdl_len: libc::c_uchar,
     pub sdl_family: libc::c_uchar,
```

```
lo0 Some(00:00:00:00:00:00)
gif0 Some(00:00:00:00:00:00)
stf0 Some(00:00:00:00:00:00)
anpi1 Some(9e:da:6a:e3:62:4b)
anpi0 Some(9e:da:6a:e3:62:4a)
en3 Some(9e:da:6a:e3:62:2a)
en4 Some(9e:da:6a:e3:62:2b)
en1 Some(36:a5:bd:13:d5:00)
en2 Some(36:a5:bd:13:d5:04)
bridge0 Some(36:a5:bd:13:d5:00)
ap1 Some(72:ed:3c:15:3d:4d)
en0 Some(50:ed:3c:15:3d:4d)
awdl0 Some(3e:6f:20:ab:35:7a)
llw0 Some(3e:6f:20:ab:35:7a)
utun0 Some(00:00:00:00:00:00)
utun1 Some(00:00:00:00:00:00)
utun2 Some(00:00:00:00:00:00)
utun4 Some(00:00:00:00:00:00)
utun5 Some(00:00:00:00:00:00)
utun3 Some(00:00:00:00:00:00)
```

**Comment by maxmouchet** [comment]
That does it, thanks! I can open a PR later on.

```diff
diff --git a/pnet_datalink/src/bindings/bpf.rs b/pnet_datalink/src/bindings/bpf.rs
index 9a938a4..81cfaa0 100644
--- a/pnet_datalink/src/bindings/bpf.rs
+++ b/pnet_datalink/src/bindings/bpf.rs
@@ -95,6 +95,7 @@ pub struct ifreq {
     target_os = "macos",
     target_os = "ios"
 ))]
+#[repr(C)]
 pub struct sockaddr_dl {
     pub sdl_len: libc::c_uchar,
     pub sdl_family: libc::c_uchar,
```

I've anonymized the MAC address but they are correct now:

```
lo0 Some(00:00:00:00:00:00)
gif0 Some(00:00:00:00:00:00)
stf0 Some(00:00:00:00:00:00)
anpi1 Some(9e:da:xx:xx:xx:xx)
anpi0 Some(9e:da:xx:xx:xx:xx)
en3 Some(9e:da:xx:xx:xx:xx)
en4 Some(9e:da:xx:xx:xx:xx)
en1 Some(36:a5:xx:xx:xx:xx)
en2 Some(36:a5:xx:xx:xx:xx)
bridge0 Some(36:a5:xx:xx:xx:xx)
ap1 Some(72:ed:xx:xx:xx:xx)
en0 Some(50:ed:xx:xx:xx:xx)
awdl0 Some(3e:6f:xx:xx:xx:xx)
llw0 Some(3e:6f:xx:xx:xx:xx)
utun0 Some(00:00:00:00:00:00)
utun1 Some(00:00:00:00:00:00)
utun2 Some(00:00:00:00:00:00)
utun4 Some(00:00:00:00:00:00)
utun5 Some(00:00:00:00:00:00)
utun3 Some(00:00:00:00:00:00)
```

**Comment by mrmonday** [comment]
Looks like a lot of structs are missing this - I'll get it fixed up now.

**Comment by mrmonday** [comment]
This should be fixed in v0.32.0 - could you please check?

**Comment by maxmouchet** [comment]
It works, thanks a lot! :)
